graveler/committed/tree/diff_test.go

```diff
@@ -0,0 +1,350 @@
+			expectedDiffKeys:         []string{},
+			expectedDiffTypes:        []graveler.DiffType{},
+			expectedDiffIdentities:   []string{},
+			expectedLeftReadsByPart:  []int{0, 0, 0, 0, 0, 0},
+			expectedRightReadsByPart: []int{0, 0, 0, 0},
+		},
+		"part added in the middle": {
+			leftKeys:                 [][]string{{"k1", "k2"}, {"k5", "k6"}},
+			leftIdentities:           [][]string{{"i1", "i2"}, {"i5", "i6"}},
+			rightKeys:                [][]string{{"k1", "k2"}, {"k3", "k4"}, {"k5", "k6"}},
+			rightIdentities:          [][]string{{"i1", "i2"}, {"i3", "i4"}, {"i5", "i6"}},
+			expectedDiffKeys:         []string{"k3", "k4"},
+			expectedDiffTypes:        []graveler.DiffType{added, added},
+			expectedDiffIdentities:   []string{"i3", "i4"},
+			expectedLeftReadsByPart:  []int{0, 0},
+			expectedRightReadsByPart: []int{0, 2, 0},
+		},
+	}
+	for name, tst := range tests {
+		t.Run(name, func(t *testing.T) {
+			fakeLeft := newFakeTreeIterator(tst.leftKeys, tst.leftIdentities)
+			fakeRight := newFakeTreeIterator(tst.rightKeys, tst.rightIdentities)
+			it := tree.NewDiffIterator(fakeLeft, fakeRight)
+			var diffs []*graveler.Diff
+			actualDiffKeys := make([]string, 0)
+			for it.Next() {
+				actualDiffKeys = append(actualDiffKeys, string(it.Value().Key))
+				diffs = append(diffs, it.Value())
+			}
+			if it.Err() != nil {
+				t.Fatalf("got unexpected error: %v", it.Err())
+			}
+			if diff := deep.Equal(tst.expectedDiffKeys, actualDiffKeys); diff != nil {
+				t.Fatalf("keys in diff different than expected. diff=%s", diff)
+			}
+			for i, d := range diffs {
+				if d.Type != tst.expectedDiffTypes[i] {
+					t.Fatalf("unexpected key in diff index %d. expected=%s, got=%s", i, tst.expectedDiffKeys[i], string(d.Key))
+				}
+				if string(d.Value.Identity) != tst.expectedDiffIdentities[i] {
+					t.Fatalf("unexpected identity in diff index %d. expected=%s, got=%s", i, tst.expectedDiffIdentities[i], string(d.Value.Identity))
+				}
+			}
+			//if diff := deep.Equal(tst.expectedLeftReadsByPart, fakeLeft.readsByPart); diff != nil {
+			//	t.Fatalf("unexpected number of reads on left parts. diff=%s", diff)
+			//}
+			//if diff := deep.Equal(tst.expectedRightReadsByPart, fakeRight.readsByPart); diff != nil {
+			//	t.Fatalf("unexpected number of reads on right parts. diff=%s", diff)
+			//}
+		})
+	}
+}
+
+func TestDiffSeek(t *testing.T) {
+	const (
+		added   = graveler.DiffTypeAdded
+		removed = graveler.DiffTypeRemoved
+		changed = graveler.DiffTypeChanged
+	)
+	left := [][]string{{"k1", "k2"}, {"k4", "k5"}, {"k6", "k7"}}
+	right := [][]string{{"k1", "k3"}, {"k4", "k5"}, {"k6", "k7"}}
+	leftIdentities := [][]string{{"i1", "i2"}, {"i4", "i5"}, {"i6", "i7"}}
+	rightIdentities := [][]string{{"i1", "i3"}, {"i4", "i5"}, {"i6", "i7a"}}
+	diffTypeByKey := map[string]graveler.DiffType{"k2": removed, "k3": added, "k7": changed}
+	diffIdentityByKey := map[string]string{"k2": "i2", "k3": "i3", "k7": "i7a"}
+
+	it := tree.NewDiffIterator(
```

Comment:
defer close

Reply:
@nopcoder  we keep on writing these leaks, I do it all the time.  Is there any way to find missing `Close`s automatically?  E.g. in C++ I would just put a debug-mode assertion in the iterator destructor, and then tools such as address-sanitizer (or even much weaker tools) would finish the job for me.  `go vet` knows how to check for unused cancel functions returned from contexts,
>     lostcancel   check cancel func returned by context.WithCancel is called

It's not fair that just because we have garbage collection we cannot have nice tools :angry: